product-mini/platforms/esp-idf/main.c

```diff
@@ -2,38 +2,109 @@
+{
+    const char *exception;
 
-extern void
-iwasm_main(void);
+    wasm_application_execute_main(module_inst, 0, NULL);
+    if ((exception = wasm_runtime_get_exception(module_inst)))
+        printf("%s\n", exception);
+    return NULL;
+}
 
-static void
-wamr_task(void *param)
+void *
+iwasm_main(void *arg)
 {
-    os_printf("WAMR task started\n");
+    (void)arg; /* unused */
+
+    printf("about to set up the stuff\n");
+
+    // setup variables for instantiating and running the wasm module
+    uint8_t *wasm_file_buf = NULL;
+    unsigned wasm_file_buf_size = 0;
+    wasm_module_t wasm_module = NULL;
+    wasm_module_inst_t wasm_module_inst = NULL;
+    char error_buf[128];
 
-    iwasm_main();
+    // configure memory allocation
+    RuntimeInitArgs init_args;
+    memset(&init_args, 0, sizeof(RuntimeInitArgs));
 
-    while (1) {
-        task_sleep(1000);
-        /*os_printf("Hello WAMR\n");*/
+    init_args.mem_alloc_type = Alloc_With_Allocator;
+    init_args.mem_alloc_option.allocator.malloc_func = (void *)os_malloc;
+    init_args.mem_alloc_option.allocator.realloc_func = (void *)os_realloc;
+    init_args.mem_alloc_option.allocator.free_func = (void *)os_free;
+
+    printf("wasm_runtime_full_init\n");
+    // initialize runtime environment
+    if (!wasm_runtime_full_init(&init_args)) {
+        printf("Init runtime failed.\n");
+        return NULL;
     }
-    (void)param;
-}
 
-static bool
-app_init(uint32_t id)
-{
-    os_printf("WAMR init, id: %d\n", id);
-    task_start("wamr_task", 8192, 4, wamr_task, NULL);
-    return true;
+    // load WASM byte buffer from byte buffer of include file
+    printf("use an internal test file, that's going to output Hello World\n");
+    wasm_file_buf = (uint8_t *)wasm_test_file;
+    wasm_file_buf_size = sizeof(wasm_test_file);
+
+    // load WASM module
+    if (!(wasm_module = wasm_runtime_load(wasm_file_buf, wasm_file_buf_size,
+                                          error_buf, sizeof(error_buf)))) {
+        printf("Error in wasm_runtime_load: %s\n", error_buf);
+        goto fail2;
+    }
+
+    printf("about to call wasm_runtime_instantiate\n");
+    if (!(wasm_module_inst =
+              wasm_runtime_instantiate(wasm_module, 32 * 1024, // stack size
+                                       32 * 1024,              // heap size
+                                       error_buf, sizeof(error_buf)))) {
+        printf("Error while instantiating: %s\n", error_buf);
+        goto fail1;
+    }
+
+    printf("run main() of the application\n");
+    void *ret = app_instance_main(wasm_module_inst);
+    assert(!ret);
+
+fail1:
+    // destroy the module instance
+    printf("wasm_runtime_deinstantiate\n");
+    wasm_runtime_deinstantiate(wasm_module_inst);
+
+fail2:
+    // unload the module
+    printf("wasm_runtime_unload\n");
+    wasm_runtime_unload(wasm_module);
+
+    // destroy runtime environment
+    printf("wasm_runtime_destroy\n");
+    wasm_runtime_destroy();
```

Comment:
Seems had better to be:
```C
    // destroy the module instance
    printf("wasm_runtime_deinstantiate\n");
    wasm_runtime_deinstantiate(wasm_module_inst);

fail2:
    // unload the module
    printf("wasm_runtime_unload\n");
    wasm_runtime_unload(wasm_module);

fail1:
    // destroy runtime environment
    printf("wasm_runtime_destroy\n");
    wasm_runtime_destroy();
```

And in line 64, `goto fail1`, in line 73, `goto fail2`